tikv/raftstore/snap.go

```diff
@@ -0,0 +1,834 @@
+		stat *SnapStatistics, deleter SnapshotDeleter) error
+	Path() string
+	Exists() bool
+	Delete()
+	Meta() (os.FileInfo, error)
+	TotalSize() uint64
+	Save() error
+	Apply(option ApplyOptions) error
+}
+
+// copySnapshot is a helper function to copy snapshot.
+// Only used in tests.
+func copySnapshot(to, from Snapshot) error {
+	if !to.Exists() {
+		_, err := io.Copy(to, from)
+		return errors.WithStack(err)
+	}
+	return nil
+}
+
+// `SnapshotDeleter` is a trait for deleting snapshot.
+// It's used to ensure that the snapshot deletion happens under the protection of locking
+// to avoid race case for concurrent read/write.
+type SnapshotDeleter interface {
+	// DeleteSnapshot returns true if it successfully delete the specified snapshot.
+	DeleteSnapshot(key *SnapKey, snapshot Snapshot, checkEntry bool) bool
+}
+
+func retryDeleteSnapshot(deleter SnapshotDeleter, key *SnapKey, snap Snapshot) bool {
+	for i := 0; i < deleteRetryMaxTime; i++ {
+		if deleter.DeleteSnapshot(key, snap, true) {
+			return true
+		}
+		time.Sleep(deleteRetryDuration)
+	}
+	return false
+}
+
+func genSnapshotMeta(cfFiles []*CFFile) (*rspb.SnapshotMeta, error) {
+	cfMetas := make([]*rspb.SnapshotCFFile, 0, len(snapshotCFs))
+	for _, cfFile := range cfFiles {
+		var found bool
+		for _, snapCF := range snapshotCFs {
+			if snapCF == cfFile.CF {
+				found = true
+			}
+		}
+		if !found {
+			return nil, errors.Errorf("failed to encode invalid snapshot CF %s", cfFile.CF)
+		}
+		cfMeta := &rspb.SnapshotCFFile{
+			Cf:       cfFile.CF,
+			Size_:    cfFile.Size,
+			Checksum: cfFile.Checksum,
+		}
+		cfMetas = append(cfMetas, cfMeta)
+	}
+	return &rspb.SnapshotMeta{
+		CfFiles: cfMetas,
+	}, nil
+}
+
+func checkFileSize(path string, expectedSize uint64) error {
+	size, err := util.GetFileSize(path)
+	if err != nil {
+		return err
+	}
+	if size != expectedSize {
+		return errors.Errorf("invalid size %d for snapshot cf file %s, expected %d", size, path, expectedSize)
+	}
+	return nil
+}
+
+func checkFileChecksum(path string, expectedChecksum uint32) error {
+	checksum, err := util.CalcCRC32(path)
+	if err != nil {
+		return err
+	}
+	if checksum != expectedChecksum {
+		return errors.Errorf("invalid checksum %d for snapshot cf file %s, expected %d",
+			checksum, path, expectedChecksum)
+	}
+	return nil
+}
+
+func checkFileSizeAndChecksum(path string, expectedSize uint64, expectedChecksum uint32) error {
+	err := checkFileSize(path, expectedSize)
+	if err == nil {
+		err = checkFileChecksum(path, expectedChecksum)
+	}
+	return err
+}
+
+type CFFile struct {
+	CF          CFName
+	Path        string
+	TmpPath     string
+	ClonePath   string
+	SstWriter   *rocksdb.SstFileWriter
+	File        *os.File
+	KVCount     int
+	Size        uint64
+	WrittenSize uint64
+	Checksum    uint32
+	WriteDigest hash.Hash32
+}
+
+type MetaFile struct {
+	Meta *rspb.SnapshotMeta
+	Path string
+	File *os.File
+
+	// for writing snapshot
+	TmpPath string
+}
+
+var _ Snapshot = new(Snap)
```

Comment:
What it the purpose for this anonymous variable.

Reply:
To make compiler check that the `Snap` implements `Snapshot` interface.